DAS-2292: regression test for the smap-l2-gridder

flamingbear opened the pull request:
## Description

This PR adds regression tests for the harmony-smap-l2-gridder.

## Jira Issue ID

[DAS-2292](https://bugs.earthdata.nasa.gov/browse/DAS-2292)


## Local Test Steps

Clone this branch.

build a local conda environment from the env file
smap-l2-gridder/environment.yml

```
❯ conda env create -f smap-l2-gridder/environment.yaml
```


Fire up a jupyter notebook and point the `harmony_host_url` at SIT.

```
harmony_host_url = 'https://harmony.sit.earthdata.nasa.gov'
```

Run the notebook and verify the tests pass.

Additionally, you can build the image locally and run under docker  from the test dir:

``` shell
❯ make smap-l2-gridder-image
❯ ./run_notebooks.sh smap-l2-gridder
```




## PR Acceptance Checklist
* [X] Acceptance criteria met
* [X] Tests added/updated (if needed) and passing
* [X] Documentation updated (if needed)
* [X] CHANGELOG updated with the changes for this PR

flamingbear commented on config/services_tests_config_prod.json:
```diff
@@ -1,22 +1,23 @@
+  "podaac-concise": "TBD",
+  "podaac-l2-subsetter": "TBD",
+  "query-cmr": "geoloco",
+  "smap-l2-gridder": "smap-l2-gridder",
```
Indentation but this line

flamingbear commented on config/services_tests_config_uat.json:
```diff
@@ -1,22 +1,23 @@
+  "podaac-concise": "TBD",
+  "podaac-l2-subsetter": "TBD",
+  "query-cmr": "variable-subsetter",
+  "smap-l2-gridder": "smap-l2-gridder",
```
same indentation

vutrannasa commented:
Test: **PASS**
- Clone branch
- Create `conda env create -f smap-l2-gridder/environment.yaml`
- Ran test/smap-l2-gridder/smap-l2-gridder_Regression.ipynb

`2228910_SMAP_L2_SM_P_E_00872_D_20150401T044811_R19240_001_regridded.nc
Downloaded: 2228910_SMAP_L2_SM_P_E_00872_D_20150401T044811_R19240_001_regridded.nc
Saved output to: [/var/folders/sx/4dq8dzl54g3gd58k8k9thd6r0000gp/T/tmpovgtfijl/SPL2SMP_E.nc](https://file+.vscode-resource.vscode-cdn.net/var/folders/sx/4dq8dzl54g3gd58k8k9thd6r0000gp/T/tmpovgtfijl/SPL2SMP_E.nc)
Success: SPL2SMP_E Test.
2228912_SMAP_L2_SM_AP_01058_D_20150413T220114_R13080_001_regridded.nc
Downloaded: 2228912_SMAP_L2_SM_AP_01058_D_20150413T220114_R13080_001_regridded.nc
Saved output to: [/var/folders/sx/4dq8dzl54g3gd58k8k9thd6r0000gp/T/tmpovgtfijl/SPL2SMAP.nc](https://file+.vscode-resource.vscode-cdn.net/var/folders/sx/4dq8dzl54g3gd58k8k9thd6r0000gp/T/tmpovgtfijl/SPL2SMAP.nc)
Success: SPL2SMAP Test.
Success: Entire Test Suite.`

flamingbear commented:
Thanks @vutrannasa, I found an issue with the github actions in my fork, and I'm going to put this into draft status again.  But your review will still verify that the tests pass as expected.

flamingbear commented:
I fixed a few items in the .github/workflow (I used smap-l2-subsetter instead of smap-l2-gridder in a few places)

Also I removed a hard coded nasa so let me run the tests in my fork.

You can test the image by pulling from my fork (if you have an intel machine)

```
❯ docker pull --platform linux/amd64 ghcr.io/flamingbear/regression-tests-smap-l2-gridder:latest
❯ export REGRESSION_TESTS_SMAP_L2_GRIDDER_IMAGE="ghcr.io/flamingbear/regression-tests-smap-l2-gridder:latest"
❯ ./run_notebooks.sh smap-l2-gridder
```

If you don't you can run the tests in the flamingbear fork against UAT:

https://github.com/flamingbear/harmony-regression-tests/actions/runs/13118824191/job/36599699023

flamingbear commented:
Welp, I didn't mean to include all of the commits I was flinging at my fork to get this to work, but there they are. 
6eb614ec39e73857eff5e44ba64f2a249c3602e0 is the commit that fixes #117 for running the test.

This was all flung at my fork to test this PR against uat.  It is "working" there but the UAT harmony has not had it's memory upgraded, so is now failing for that reason.

flamingbear commented on .github/workflows/notebook-test-suite.yml:
```diff
@@ -133,7 +134,11 @@ jobs:
 
       - name: Pull Docker image
-        run: docker pull ghcr.io/nasa/regression-tests-${{ matrix.service }}:${{ github.event.client_payload.docker-image-tag || inputs.docker-image-tag || 'latest' }}
+        run: docker pull ghcr.io/${{github.repository_owner}}/regression-tests-${{ matrix.service }}:${{ github.event.client_payload.docker-image-tag || inputs.docker-image-tag || 'latest' }}
+
+      - name: Update Test Image Name
+        run: |
+          echo "REGRESSION_TESTS_$(echo "${{ matrix.service }}" | tr '[:lower:]' '[:upper:]' | tr '-' '_')_IMAGE=ghcr.io/${{ github.repository_owner }}/regression-tests-${{ matrix.service }}:${{ github.event.client_payload.docker-image-tag || inputs.docker-image-tag || 'latest' }}" >> $GITHUB_ENV
```
This is a change I added to get a forked repo to be able to run the tests.  It's awkward, but I don't see other simple options.

flamingbear replied:
Not fully correct.  Makefile has nothing to do with the .github actions.
This is updating to remove the hard coded nasa from the docker pull via ${{github.repository_owner}} on L137

Then L141 is setting the REGRESSION_TESTS_[testname fixedup]_IMAGE to the value of the docker image that is desired/just built.

When the environment variable is set it is used in preference to the default image. Which is how you can choose which image should be used as opposed to the default.  now we're just setting this for each run, to the default value on the nasa github or your correct value in your fork.

flamingbear commented:
Last comments https://github.com/flamingbear/harmony-regression-tests/actions/runs/13120170428/job/36604039292 is an example of running a test in a fork with a specific version 0.0.1 (even though it fails as expected)

owenlittlejohns reviewed (commented):
I ran the tests locally (against Harmony SIT in two ways):

* ✅ Creating the conda environment and running the notebook via a notebook server. That all worked!
* ❗ Building the image with the `Makefile` and using `run_notebooks.sh`. This had a `DeadKernel` error. @flamingbear and I talked about this a little bit and it sounds like an architecture issue, but I'll let @flamingbear weigh in with more details.

Everything else looked good except for some small nits.

flamingbear commented on test/shared_utils/compare.py:
```diff
@@ -27,8 +27,8 @@ def compare_results_to_reference_file(
 
-    reference_groups = open_groups(reference_file_name)
-    results_groups = open_groups(results_file_name)
+    reference_groups = open_groups(reference_file_name, decode_timedelta=False)
+    results_groups = open_groups(results_file_name, decode_timedelta=False)
```
No, this is in response to a future warning 
```
FutureWarning: In a future version of xarray decode_timedelta will default to False rather than None. To silence this warning, set decode_timedelta to True, False, or a 'CFTimedeltaCoder' instance.
```
setting it explicitly to `False` from the previous `None`, but just silences the warning.

owenlittlejohns replied:
Thanks for the clarification.

flamingbear commented on test/smap-l2-gridder/smap-l2-gridder_Regression.ipynb:
```diff
@@ -0,0 +1,260 @@
+    "## Regression tests for the Harmony SMAP L2 Gridder \n",
+    "\n",
+    "\n",
+    "This Jupyter notebook runs a suite of regression tests against some requests against the Harmony SMAP L2 gridding service. \n",
```
07267a5

flamingbear commented on test/smap-l2-gridder/smap-l2-gridder_Regression.ipynb:
```diff
@@ -0,0 +1,260 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with TemporaryDirectory() as tmp_dir:\n",
+    "    if configuration is not None:\n",
+    "        for collection, test_config in configuration['config'].items():\n",
```
It doesn't matter.

flamingbear replied:
but I changed it anyway

flamingbear commented on test/smap-l2-gridder/smap-l2-gridder_Regression.ipynb:
```diff
@@ -0,0 +1,260 @@
+    "* SIT: <https://harmony.sit.earthdata.nasa.gov>\n",
+    "* Local: <http://localhost:3000>\n",
+    "\n",
+    "The default value is for the UAT environment. When using this notebook there are two ways to use the non-default environment:\n",
```
Yes, I checked in the wrong value, it should be fixed now.

flamingbear replied:
83e64c0f6418dcf76af0a0aad93e8fd1ab63b67c

flamingbear commented:
There is a notebook attached to the DAS-2292 ticket showing the subsetting in action (via how the reference files were generated).

owenlittlejohns commented on README.md:
```diff
@@ -243,7 +243,7 @@ dependencies:
 ### Versioning
 
-The regression test notebooks follow semantic versioning:
+The regression test notebooks try to follow semantic versioning:
```
https://giphy.com/gifs/starwars-star-wars-yoda-26FmQ6EOvLxp6cWyY

(I get it, though, it's not completely straightforward to apply semantic versioning to these notebooks)

owenlittlejohns commented on test/shared_utils/compare.py:
```diff
@@ -18,17 +18,34 @@ def compare_results_to_reference_file(
-    """Use `DataTree` functionality to compare data values, variables,
-    coordinates, metadata, and all their corresponding attributes of
-    downloaded results to a reference file.
+    """Compare two files as DataTrees
```
Nice improvement to the documentation string!

owenlittlejohns commented on test/shared_utils/compare.py:
```diff
@@ -38,6 +55,10 @@ def compare_results_to_reference_file(
 
+    # Limit comparison of data
+    if subset_selector is not None:
+        results_data = subset_datatree(results_data, subset_selector)
```
Okay, again just asking basic things for clarification:

In the instance that `subset_selector` is specified, you're saying the output results file needs to be subsetted, _but_ that it is assumed that the reference file already has that subsetting applied to it.

If that's the case, it might be worth mentioning that in the documentation string. (i.e. that the reference file is expected to already have such subsetting applied)

flamingbear replied:
Yes, I could leave the entire datatree in the reference file and subset both, but they are just a little bit smaller if they're stored subsetted.

flamingbear replied:
1a90cc2

owenlittlejohns commented on test/smap-l2-gridder/smap-l2-gridder_Regression.ipynb:
```diff
@@ -0,0 +1,298 @@
+    "\n",
+    "* The cell below loops through the SMAP L2 collections for the configured environment.\n",
+    "* A request for gridding the L2 data is submitted to Harmony and the results downloaded.\n",
+    "* Verification compares the results to the a previously generated reference data file.  This comparison is done only on the first 2000 rows and 2000 columns of the downloaded data to save memory usage. \n"
```
Thanks for adding these details - it helps.

Super nit: "...compares the results to ~~the~~ a previously generated..."

flamingbear replied:
[2bfc92e](https://github.com/nasa/harmony-regression-tests/pull/130/commits/2bfc92eccae724b672e1600a3649d56d7ed35b09)

owenlittlejohns reviewed (approved):
I ran the tests locally via the `Makefile` and `run_notebooks.sh` and this time the Docker container didn't run out of memory, so I think your fix to limit the size of the arrays being compared worked nicely.

I only have nits left at this point - thanks for addressing a bunch of them already.

```
(papermill-net2cog) OwenLittlejohns@GSLAL0324050268:~/Documents/git/NASA/harmony-regression-tests/test$ ./run_notebooks.sh smap-l2-gridder
Running regression tests
Test suite smap-l2-gridder starting
running test with ghcr.io/nasa/regression-tests-smap-l2-gridder:latest
Waiting for smap-l2-gridder.
Input Notebook:  smap-l2-gridder/smap-l2-gridder_Regression.ipynb
Output Notebook: /workdir/output/smap-l2-gridder/Results.ipynb
Working directory: smap-l2-gridder
Executing:   0%|          | 0/13 [00:00<?, ?cell/s]Executing notebook with kernel: python3
Executing: 100%|██████████| 13/13 [02:09<00:00,  9.96s/cell]
Test suite smap-l2-gridder succeeded
Tests completed (passed)
```

owenlittlejohns commented on test/shared_utils/compare.py:
```diff
@@ -18,17 +18,34 @@ def compare_results_to_reference_file(
+        coordinates_to_fix: List of coordinate names to be renamed in the case
+                            that the input has "unalignable" names.
+
+        subset_selector: Dict of top level group names to selection
+                         dictionaries used to subset the input DataTree to the
+                         reference data. (see subset_datatree)
```
I looked a bit more at where it's called, and "selection dictionaries" does make sense after all. But an example in the doc-string would still help a bit.

flamingbear replied:
Did you "see subset_datatree" that's where it's being used and defined. would it help to move that up before this function in the file? I don't want to copy and paste the example

flamingbear replied:
1a90cc2
